### ISSUE TITLE
What's wrong with my code from dgl to pyg？

### ISSUE THREAD

**Issue by tjb-tech** [pyg-team/pytorch_geometric#7017]
### 🐛 Describe the bug

I am transforming a dgl version code to pyg version code. The dgl one build a GraphSAGE model for transductive and inductive node classification on dataset Cora. So I aslo transform these two setting using pyg. I observe that the transductive setting using two library are close. The following codes show the data loader under two library in transductive setting: 
pyg
```python
idx_train, idx_val, idx_test = indices

    feats = feats.to(device)
    labels = labels.to(device)

    if "SAGE" in model.model_name:
        # Create dataloader for SAGE

        # Create csr/coo/csc formats before launching sampling processes
        # This avoids creating certain formats in each data loader process, which saves memory and CPU.
        

        dataloader = NeighborLoader(pyg_data, [eval(fanout) for fanout in conf["fan_out"].split(",")], idx_train, 
                        batch_size=batch_size, shuffle=True, drop_last=False, num_workers=conf["num_workers"])
        

        # SAGE inference is implemented as layer by layer, so the full-neighbor sampler only collects one-hop neighors
        
        dataloader_eval = NeighborLoader(pyg_data, [-1], torch.arange(pyg_data.num_nodes), 
                        batch_size=batch_size, shuffle=False, drop_last=False, num_workers=conf["num_workers"])

        
        # Add global node index information.
        dataloader_eval.data.num_nodes = pyg_data.num_nodes
        dataloader_eval.data.n_id = torch.arange(pyg_data.num_nodes)
```
dgl
```python
idx_train, idx_val, idx_test = indices

    feats = feats.to(device)
    labels = labels.to(device)

    if "SAGE" in model.model_name:
        # Create dataloader for SAGE

        # Create csr/coo/csc formats before launching sampling processes
        # This avoids creating certain formats in each data loader process, which saves memory and CPU.
        g.create_formats_()
        sampler = dgl.dataloading.MultiLayerNeighborSampler(
            [eval(fanout) for fanout in conf["fan_out"].split(",")]
        )
        dataloader = dgl.dataloading.DataLoader(
            g,
            idx_train,
            sampler,
            batch_size=batch_size,
            shuffle=True,
            drop_last=False,
            num_workers=conf["num_workers"],
        )

        # SAGE inference is implemented as layer by layer, so the full-neighbor sampler only collects one-hop neighors
        sampler_eval = dgl.dataloading.MultiLayerFullNeighborSampler(1)
        dataloader_eval = dgl.dataloading.DataLoader(
            g,
            torch.arange(g.num_nodes()),
            sampler_eval,
            batch_size=batch_size,
            shuffle=False,
            drop_last=False,
            num_workers=conf["num_workers"],
        )
```
However, for inductive setting, the results using pyg are much worse than those using dgl. The model hyperparameters and other settings are the same as transductive setting. The only difference is the dataloader, which is shown as below: 
pyg
```python
obs_idx_train, obs_idx_val, obs_idx_test, idx_obs, idx_test_ind = indices

    feats = feats.to(device)
    labels = labels.to(device)
    obs_feats = feats[idx_obs]
    obs_labels = labels[idx_obs]

    obs_pyg_data = pyg_data.subgraph(idx_obs)
    
    obs_pyg_data.x = obs_feats
    obs_pyg_data.y = obs_labels
    obs_pyg_data.num_edges = obs_pyg_data.edge_index.shape[1]
    print(obs_pyg_data, len(idx_obs))


    if "SAGE" in model.model_name:
        # Create dataloader for SAGE

        # Create csr/coo/csc formats before launching sampling processes
        # This avoids creating certain formats in each data loader process, which saves momory and CPU.
        
        obs_dataloader = NeighborLoader(copy.deepcopy(obs_pyg_data), [eval(fanout) for fanout in conf["fan_out"].split(",")], obs_idx_train, 
                        batch_size=batch_size, shuffle=True, drop_last=False, num_workers=conf["num_workers"])

        
        obs_dataloader_eval = NeighborLoader(copy.deepcopy(obs_pyg_data), [-1], torch.arange(obs_pyg_data.x.shape[0]), 
                        batch_size=batch_size, shuffle=False, drop_last=False, num_workers=conf["num_workers"])
        del obs_dataloader_eval.data.x, obs_dataloader_eval.data.y
        obs_dataloader_eval.data.num_nodes = obs_pyg_data.x.shape[0]
        obs_dataloader_eval.data.n_id = torch.arange(obs_pyg_data.x.shape[0])

        dataloader_eval = NeighborLoader(copy.deepcopy(pyg_data), [-1], torch.arange(pyg_data.num_nodes), 
                        batch_size=batch_size, shuffle=False, drop_last=False, num_workers=conf["num_workers"])
        del dataloader_eval.data.x, dataloader_eval.data.y
        dataloader_eval.data.num_nodes = pyg_data.num_nodes
        dataloader_eval.data.n_id = torch.arange(pyg_data.num_nodes)
```
dgl
```python
obs_idx_train, obs_idx_val, obs_idx_test, idx_obs, idx_test_ind = indices

    feats = feats.to(device)
    labels = labels.to(device)
    obs_feats = feats[idx_obs]
    obs_labels = labels[idx_obs]
    obs_g = g.subgraph(idx_obs)

    print(len(obs_g.edges()[0]))

    if "SAGE" in model.model_name:
        # Create dataloader for SAGE

        # Create csr/coo/csc formats before launching sampling processes
        # This avoids creating certain formats in each data loader process, which saves momory and CPU.
        obs_g.create_formats_()
        g.create_formats_()
        sampler = dgl.dataloading.MultiLayerNeighborSampler(
            [eval(fanout) for fanout in conf["fan_out"].split(",")]
        )
        obs_dataloader = dgl.dataloading.DataLoader(
            obs_g,
            obs_idx_train,
            sampler,
            batch_size=batch_size,
            shuffle=True,
            drop_last=False,
            num_workers=conf["num_workers"],
        )

        sampler_eval = dgl.dataloading.MultiLayerFullNeighborSampler(1)
        obs_dataloader_eval = dgl.dataloading.DataLoader(
            obs_g,
            torch.arange(obs_g.num_nodes()),
            sampler_eval,
            batch_size=batch_size,
            shuffle=False,
            drop_last=False,
            num_workers=conf["num_workers"],
        )
        dataloader_eval = dgl.dataloading.DataLoader(
            g,
            torch.arange(g.num_nodes()),
            sampler_eval,
            batch_size=batch_size,
            shuffle=False,
            drop_last=False,
            num_workers=conf["num_workers"],
        )
```
the model is the standard GraphSAGE, which is shown as follow: 
```python
class SAGE(nn.Module):
    def __init__(
            self,
            num_layers,
            input_dim,
            hidden_dim,
            output_dim,
            dropout_ratio,
            activation,
            norm_type="none",
    ):
        super().__init__()
        self.num_layers = num_layers
        self.hidden_dim = hidden_dim
        self.output_dim = output_dim
        self.norm_type = norm_type
        self.activation = activation
        self.dropout = nn.Dropout(dropout_ratio)
        self.layers = nn.ModuleList()
        self.norms = nn.ModuleList()

        if num_layers == 1:
            self.layers.append(SAGEConv(input_dim, output_dim))
        else:
            self.layers.append(SAGEConv(input_dim, hidden_dim))
            if self.norm_type == "batch":
                self.norms.append(nn.BatchNorm1d(hidden_dim))
            elif self.norm_type == "layer":
                self.norms.append(nn.LayerNorm(hidden_dim))

            for i in range(num_layers - 2):
                self.layers.append(SAGEConv(hidden_dim, hidden_dim))
                if self.norm_type == "batch":
                    self.norms.append(nn.BatchNorm1d(hidden_dim))
                elif self.norm_type == "layer":
                    self.norms.append(nn.LayerNorm(hidden_dim))

            self.layers.append(SAGEConv(hidden_dim, output_dim))

    def forward(self, x, edge_index):
        h = x
        h_list = []
        print('training, input:', x.shape)
        for l, layer in enumerate(self.layers):
            
            h = layer(h, edge_index)
            if l != self.num_layers - 1:
                h_list.append(h)
                if self.norm_type != "none":
                    h = self.norms[l](h)
                h = self.activation(h)
                h = self.dropout(h)
            
        print(h.shape)
        return h

    def inference(self, x_all, subgraph_loader):
        """
        Inference with the GraphSAGE model on full neighbors (i.e. without neighbor sampling).
        dataloader : The entire graph loaded in blocks with full neighbors for each node.
        feats : The input feats of entire node set.
        """
        device = x_all.device
        emb_list = []
        pbar = tqdm(total=len(subgraph_loader.dataset) * len(self.layers))
        pbar.set_description('Evaluating')
        
        for l, layer in enumerate(self.layers):
            # add-on
            xs = []
            hs = []
            for batch in subgraph_loader:
                x = x_all[batch.n_id.to(device)].to(device)
                x = layer(x, batch.edge_index.to(device))
                if l != self.num_layers - 1:
                    hs.append(x[:batch.batch_size])
                    if self.norm_type != "none":
                        x = self.norms[l](x)
                    x = self.activation(x)
                    x = self.dropout(x)
                xs.append(x[:batch.batch_size])

                pbar.update(batch.batch_size)

            x_all = torch.cat(xs, dim=0)
            if l != self.num_layers - 1:
                hidden_emb = torch.cat(hs, dim = 0)
                emb_list.append(hidden_emb)
        pbar.close()
        return x_all, emb_list
```
Could you please help me to tell me whether my transformed code related to dataloader is right? And where might the problem arise？I have tried a lot of setting for different seeds, and the accuracy of dgl version is 80+, while that of pyg is only 70-. Looking forward to your reply!

### Environment

* PyG version: 
* PyTorch version:
* OS:
* Python version:
* CUDA/cuDNN version:
* How you installed PyTorch and PyG (`conda`, `pip`, source):
* Any other relevant information (*e.g.*, version of `torch-scatter`):

**Comment by rusty1s** [comment]
Interesting. I did not run your code yet, but to me it an accuracy of 0.7 sounds more realistic in the inductive graph setting (since there only exists like 140 training nodes that are arbitrarily sampled from the full graph). To debug, I would suggest to confirm that induced subgraphs from PyG and DGL are indeed identical. Let me know if you need any help with this.

**Comment by tjb-tech** [comment]
> @rusty1s ，Thanks for your reply! I have worked on it for one day, and I found two solutions with the same subgraph (the shape of subgraph edge_index, x and y are all the same), but different results! BTW, the model setting are all the same. 
One is :
```python 
obs_idx_train, obs_idx_val, obs_idx_test, idx_obs, idx_test_ind = indices

    feats = feats.to(device)
    labels = labels.to(device)
    obs_feats = feats[idx_obs]
    obs_labels = labels[idx_obs]
    def to_inductive(data, idx_obs, feats, label):
        obs_data = data.subgraph(idx_obs)
        edge_index = obs_data.edge_index
        x = feats
        y = label
        ret = Data(x=x, edge_index=edge_index, y=y, num_nodes = feats.shape[0])
        return ret
    # 1.  
    ori_data = copy.deepcopy(pyg_data)
    obs_feats_ori = copy.deepcopy(obs_feats)
    obs_labels_ori = copy.deepcopy(obs_labels)
    obs_pyg_data = to_inductive(ori_data, idx_obs, obs_feats_ori, obs_labels_ori)
```
the score on inductive test set is only 60+. 
And the other is :
```python 
    feats = feats.to(device)
    labels = labels.to(device)
    obs_feats = feats[idx_obs]
    obs_labels = labels[idx_obs]
    def dgl_to_pyg(graph, feats, label):
        edge_index = graph.edges()
        x = feats
        y = label
        data = Data(x=x, edge_index=torch.stack(edge_index, dim = 0), y=y, num_nodes = graph.num_nodes())
        return data
    # 2. 
    obs_g = kwargs.get('g').subgraph(idx_obs)
    obs_pyg_data = dgl_to_pyg(obs_g, obs_feats, obs_labels)
    
```
the `kwargs.get('g')` here denotes the origional dgl's graph, and the score on inductive test set is only 80, which is similar to the dgl version code. 
I am very confused  about these. BTW, the `obs_pyg_data` is used as below: 
```python 
obs_dataloader = NeighborLoader(copy.deepcopy(obs_pyg_data), [eval(fanout) for fanout in conf["fan_out"].split(",")], obs_idx_train, 
                        batch_size=batch_size, shuffle=True, drop_last=False, num_workers=conf["num_workers"])

        
        obs_dataloader_eval = NeighborLoader(copy.deepcopy(obs_pyg_data), [-1], torch.arange(obs_pyg_data.x.shape[0]), 
                        batch_size=batch_size, shuffle=False, drop_last=False, num_workers=conf["num_workers"])
        del obs_dataloader_eval.data.x, obs_dataloader_eval.data.y
        obs_dataloader_eval.data.num_nodes = obs_pyg_data.x.shape[0]
        obs_dataloader_eval.data.n_id = torch.arange(obs_pyg_data.x.shape[0])
```
Could please help me to find out where the error may occur? Looking forward to your reply!

**Comment by tjb-tech** [comment]
> Interesting. I did not run your code yet, but to me it an accuracy of 0.7 sounds more realistic in the inductive graph setting (since there only exists like 140 training nodes that are arbitrarily sampled from the full graph). To debug, I would suggest to confirm that induced subgraphs from PyG and DGL are indeed identical. Let me know if you need any help with this.
To be specific, the results of different datasets in different setting used dgl, demonstrated as below: 
![image](https://user-images.githubusercontent.com/56426168/227468473-5c9ba267-ca82-4af8-bc0c-2f90a6ad9659.png)
I know that it's a little bit unrealistic that the results of transductive setting are worse than those of inductive setting, but it's true. 
If you are interested in these paper, the paper is https://arxiv.org/abs/2110.08727

**Comment by rusty1s** [comment]
Thanks. Can you post a minimal PyG example to reproduce this? I just modified `examples/gcn.py` to only train on
```python
train_data = data.subgraph(data.train_mask)
```
and the performance drops from `0.815` to `0.753`, which is exactly what I would expect :)

I know the GLNN paper, but I doubt the use the same splitting as the original GCN paper. With the original split, this results in a training data of 140 nodes, but only bi-directional 21 edges:
```
Data(x=[140, 1433], edge_index=[2, 42], y=[140], train_mask=[140], val_mask=[140], test_mask=[140])
```
so it is to be expected that the performance drops here significantly.
